File: 프로그래머스/lv3/42861. 섬 연결하기/섬 연결하기.py
def solution(n, costs):
    answer = 0
    uni = [0] * (n + 1)
    
    costs.sort(key = lambda x: x[2])
    print(costs)
    for i in range(1,n + 1):
        uni[i] = i
        
    def union(x,y):
        a = find(x)
        b = find(y)
        uni[a] = b
            
    def find(x):
        print("find",x,uni);
        if uni[x] == x: return x
        else:
            uni[x] = find(uni[x])
            return uni[x]
        
    for n1,n2,cost in costs:
        print("for",n1,n2,cost);
        if find(n1) != find(n2):
            print("union")
            union(n1,n2)
            answer += cost
    return answer

File: 프로그래머스/lv3/42861. 섬 연결하기/섬 연결하기.java
import java.util.Arrays;

class Solution {
    int[] dp;
    public int solution(int n, int[][] costs) {
        int answer = 0;
        dp = new int[n];
        for(int i = 0; i < n; i++) dp[i] = i;
        Arrays.sort(costs,(o1,o2) -> o1[2] - o2[2]);

        for(int[] cost : costs){
            int start= cost[0], end=cost[1], dist=cost[2];
            if(find(start) != find(end)){
                union(start,end);
                answer += dist;
            }
        }
        
        return answer;
    }
    public int find(int x){
        if(dp[x] == x) return dp[x];
        return find(dp[x]);
    }
    
    public void union(int x, int y){
        int start = find(x);
        int end = find(y);
        dp[start] = end;
    }
}